### PR TITLE
Add public / team stats. Fixes #429.

### LINKED ISSUE #429: Add public stats to team/public problems page
It would be great to add a subset of the now only jury available stats also to the team and public pages.

I'm thinking of the submission graph over time (but all rejected verdicts bundled into one, of course nothing after the freeze) and the total number of attempts/corrects.

### PR DESCRIPTION
**Pull request by nickygerritsen**
This moves the stats logic to a service, so we can reuse it on the team/public pages. 

On the team/public problem pages it shows this if enabled:
image

### CODE REVIEW

**Comment by thijskh** [comment]
So the stats are in the team interface but not related to the specific team?

**Comment by nickygerritsen** [comment]
Yes since there is no easy way to go to the public interface if you are logged in

**Comment by thijskh** [comment]
Is it fast enough to expose it to a larger audience? I.e., does it only depend on the scorecache tables?

File: webapp/templates/partials/analysis/contest_stats.html.twig [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+                Calculated as:<br>
+                average across all teams of [3 people x (number of minutes elapsed - minute last problem solved by team t)]
+              ">Misery
+                    index</abbr>: {{ misc.misery_index|number_format(0) }} minutes
+                <br>
+            {% endif %}
+            # Submissions: {{ misc.total_submissions }}<br>
+            # Accepted Submissions: {{ misc.total_accepted }} <span
+                class="text-muted">(~{{ (misc.total_submissions>0 ? (misc.total_accepted/misc.total_submissions)*100:0)|number_format(0) }}%)</span><br>
+            # of Teams: {{ misc.num_teams }}<br>
```
Comment (thijskh) [inline webapp/templates/partials/analysis/contest_stats.html.twig:23]: The third bullet is the only one that has the word "or" in it. Is the "#" even needed or could it just say: "Submissions: 1353"?

Comment (nickygerritsen) [reply webapp/templates/partials/analysis/contest_stats.html.twig:23]: Note that this didn't change in this PR. But I can agree we can remove the # (and the 'of')

**Comment by nickygerritsen** [comment]
> Is it fast enough to expose it to a larger audience? I.e., does it only depend on the scorecache tables?

It does not use scorecache, since that doesn't contain timing information for all but the last-or-correct submission.

However, for the NWERC data it loads in ~400 ms in debug mode on my laptop in Docker, so it seems to be not that slow.

**Comment by meisterT** [comment]
I am wondering whether we should just fold this into the problemset page instead?

**Comment by nickygerritsen** [comment]
> I am wondering whether we should just fold this into the problemset page instead?

Place it below the problems there? That might make sense

**Comment by meisterT** [comment]
Or the data for each problem in the box for that problem.

**Comment by nickygerritsen** [comment]
So only attempted / correct as numbers? And maybe correct %?
And what about the contest stats and submissions over time? Don't add these?

**Comment by meisterT** [comment]
> So only attempted / correct as numbers? And maybe correct %?

Perhaps a pie chart with absolute numbers? After the results are revealed this could change from incorrect/correct to actual verdicts.


> And what about the contest stats and submissions over time? Don't add these?

For the submissions over time: I could imagine to show a heat map like chart that shows the number of successful submissions for each problem as well.
Each problem box is right now ~300px wide (on my screen at least), so use divide this into X little boxes each ~10px wide (each representing a block of time) and show with a color gradient how many successful submissions there were in that bucket of time, similar to
image

Of course I am just thinking out loud - feel free to ignore me and go ahead with what you already had planned.

**Comment by nickygerritsen** [comment]
I do like the blocks. Do we also want to show somehow which ones (or how many) of them are correct?
For the pie charts: I'm not sure if that is useful? It probably takes more space than 2/3 numbers so what does it add?

Also thinking out loud

**Comment by meisterT** [comment]
What about two rows of blocks then -  one for correct and one for incorrect submissions? (and the raw numbers somewhere)

**Comment by nickygerritsen** [comment]
Let me try to build it, sounds like it might look good

**Comment by eldering** [comment]
> > Is it fast enough to expose it to a larger audience? I.e., does it only depend on the scorecache tables?
> 
> It does not use scorecache, since that doesn't contain timing information for all but the last-or-correct submission.
> 
> However, for the NWERC data it loads in ~400 ms in debug mode on my laptop in Docker, so it seems to be not that slow.

That sounds ok to me as long as it's on a page that doesn't get viewed a lot or has a high refresh rate: 400ms is a lot if this gets hit often in a big contest with 200+ teams, say. So problems overview page sounds fine.

**Comment by nickygerritsen** [comment]
If we do @meisterT his plan it might be that the queries are faster as well. I will check

**Comment by nickygerritsen** [comment]
Updated code, update screenshot in description to reflect new code.

**Comment by nickygerritsen** [comment]
Ah and for timing: the page loads in 250 ms while it loads in 50 ms without the stats, so the speed has increased indeed.

**Review by meisterT — APPROVED** [review]

File: webapp/src/Twig/TwigExtension.php [hunk 2]
```diff
@@ -150,6 +150,7 @@ public function getGlobals()
             'refresh_flag' => $refresh_flag,
             'icat_url' => defined('ICAT_URL') ? ICAT_URL : null,
             'external_ccs_submission_url' => $this->config->get('external_ccs_submission_url'),
+            'show_public_stats' => $this->config->get('show_public_stats'),
```
Comment (thijskh) [inline webapp/src/Twig/TwigExtension.php:153]: Why is this extra global needed? The template can just see if stats is filled?

Comment (nickygerritsen) [reply webapp/src/Twig/TwigExtension.php:153]: You are right. This was needed before when it was a separate page. I will change it

**Review by thijskh — APPROVED** [review]

**Pull request merged**